Implement Reporter System

aeslaughter opened the pull request:
See https://github.com/idaholab/moose/pull/15358

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Initial implementation of Reporter system

(refs #11323)

aeslaughter committed:
Create Reporter object of iteration information

(refs #11323)

aeslaughter committed:
Improve interaction between RepoterState and ReporterContext

(refs #11323)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/578266/) on c1e3d01 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15710/site)

This comment will be updated on new commits.

aeslaughter committed:
Remove 'auto' to support min compiler

(refs #11323)

aeslaughter committed:
Add scatter and gather context

rwcarlsen commented:
You'll want to make sure you integrate the reporter name/retrieval into func parser and such - so people can reference reporter names in those input file expressions.

aeslaughter commented:
For parallel efficiency the producer must dictate how values are created (e.g., locally, one process,...) and the consumer must be able to indicated what is needed (local value, global value,...). The parallel operations that are performed per Reporter value will then governed by the interaction between the method of production and consumption.

Here is a list of what we need to consider:

- Value produced locally and consumed locally
- Value produced locally and consumed by all processors
- Value produced locally and consumed by root process

aeslaughter committed:
Creating auto parallel and error checking for producer/consumers of reporter values

(refs #11323)

aeslaughter committed:
Add header only JSON library

https://github.com/nlohmann/json

aeslaughter committed:
JSONOutput: create basic capability for outputting to a JSON file

(refs #11323)

aeslaughter committed:
Add very basic JSON diff tool

(refs #11323)

aeslaughter committed:
Remove old JSON library

(refs #11323)

aeslaughter committed:
Update MOOSE for nlohmann::json library

(refs #11323)

aeslaughter committed:
Migrate Postprocessors to use ReporterData and correct const correctness of interfaces

(refs #11323)

aeslaughter committed:
Automaticallly output Real Reporter values to CSV and screen

(refs #11323)

aeslaughter committed:
Update Executioner to use Reporter value rather than create a PP

(refs #11323)

moosebuild commented:
Job [Test debug](https://www.moosebuild.org/job/10/) on 5434f86 : invalidated by @moosebuild

try with a lower -j when performing tests

aeslaughter committed:
Perform ReporterContext initialization on-demand

(refs #11323)

aeslaughter committed:
Use MooseEnum for producer/consumer tracking and move producer to ReporterContext

(refs #11323)

aeslaughter committed:
Convert VPP objects to use Reporter system

(refs #11323)

aeslaughter committed:
Minor cleanup of API

(refs #11323)

aeslaughter committed:
Make access to ReporterData const correct

(refs #11323)

aeslaughter committed:
Use std::list only in ReporterState

(refs #11323)

aeslaughter committed:
Remove/improve deprecated calls

(refs #11323)

aeslaughter committed:
Convert geochem module to new JSON library

(refs #11323)

aeslaughter committed:
Update fluid properties to new json package

(refs #11323)

aeslaughter committed:
Add SystemAttrib to getUserObject method

(refs #11323)

aeslaughter committed:
Update stocastic tools for changes to PP/VPP systems

(refs #11323)

aeslaughter committed:
Allow Postprocessor object to operate from UserObjects block

(refs #11323)

aeslaughter committed:
Warnings and valgrind fixes

(refs #11323)

aeslaughter committed:
Update test to work with new VPP declare methods

(refs #11323)

aeslaughter committed:
Remove ExecutionerAttributeReporter

(refs #11323)

aeslaughter committed:
Rename load/store for JSON objects to avoid overload/virtual problems

(refs #11323)

aeslaughter committed:
Create IterationInfo and start documentation of Reporter system

(refs #11323)

Commit:
Fix out of order MPI calls

The ReporterData was storing the ReporterContext objects in a set of unique_ptrs, so the ordering was different across processors causing the MPI routines to be called out of order

aeslaughter committed:
Add libmesh/petsc/slepc version to skip list in JSON diff

(refs #11323)

aeslaughter committed:
Create MeshInfo reporter

(refs #11323)

aeslaughter committed:
Improve JSON output for distributed data

(refs #11323)

aeslaughter committed:
Improve JSON output and JSONDiff to handle non-static fields

(refs #11323)

aeslaughter committed:
Sort Context objects upon creation

(refs #11323)

dschwen commented:
This has the flavor of "scalar materials" (as in scalar kernels) with "properties". Properties can be of any type and have unique(?) names. Is that far off?

dschwen commented:
Also with materials we saw a lot of hard coded names in the early years, which moved to customizable property names (with `base_name` etc.). Should we expect the same in this system?

aeslaughter commented:
The Reporter objects can declare any number of ReporterValues with arbitrary names and types. The retrieval of the values mimics the VPP system that includes the object name and the vector name. However, I updated InputParameters object to handle the ReporterName type in a single parameter: `reporter = object_name:value_name`.

It would be trivial for the Reporter to declare as in MaterialProperties. I am thinking we add another special typedef "ReporterValueName". Then the declare function will first check for a parameter with a type "ReporterValueName", if it is there use the value from the parameter if not use the string as normal. I haven't done this yet, but can add it on.

aeslaughter commented:
Also, if you have create the correct to_json function (i.e., as in DataLoad/Store) the values will also automatically output with "json=true" in the outputs block.

aeslaughter committed:
Add missing includes

(refs #11323)

aeslaughter committed:
Use namespace for hash specialization for minimum compiler support

(refs #11323)

aeslaughter committed:
Add missing gold file

(refs #11323)

aeslaughter committed:
Improve trainer error message and fix reporter mode error

(refs #11323)

aeslaughter committed:
Expand PP/VPP interfaces to allow for value/object checks

(refs #11323)

aeslaughter committed:
Add missing include

(refs #11323)

aeslaughter committed:
Add documentation for Reporters

(refs #11323)

aeslaughter committed:
Improve support for key skipping in JSON diff

(refs #11323)

aeslaughter committed:
Add news of Reporter system

(refs #11323)

aeslaughter committed:
Correct reporter mode

(refs #11323)

aeslaughter committed:
Remove test that was eliminated but showed back up after rebase

(refs #11323)

aeslaughter committed:
Add missing requirement

(refs #11323)

zachmprince commented on framework/src/parser/Parser.C:
```diff
@@ -1996,6 +2004,19 @@ Parser::setScalarParameter<PostprocessorName, PostprocessorName>(
 
+template <>
+void
+Parser::setScalarParameter<ReporterName, std::string>(
+    const std::string & full_name,
+    const std::string & /*short_name*/,
+    InputParameters::Parameter<ReporterName> * param,
+    bool /*in_global*/,
+    GlobalParamsAction * /*global_block*/)
+{
+  std::vector<std::string> names = MooseUtils::split(_root->param<std::string>(full_name), "::");
+  param->set() = ReporterName(names[0], names[1]);
+}
```
This doesn't seem to be documented anywhere. Basically to set a reporter name input parameter you have `object_name::value_name`. This confused me at first because the resulting combined name is `object_name/value_name`. I think this will be confusing for others too because the CSV and console output of a reporter is `object_name/value_name`. Ideally, I think we should choose one or the other. But I can easily be convinced that this is impossible.

zachmprince replied:
Also there needs to be an error if `::` isn't in the name. There's a segfault when `::` isn't used.

aeslaughter replied:
I changed this to a / and added some doco

zachmprince reviewed (changes requested):
This is really exciting! I think the system is beautiful. Just have comments to clean some things up. I will have to start coding my own reporter to really understand and test it.

aeslaughter commented on framework/doc/content/syntax/Reporters/index.md:
```diff
@@ -0,0 +1,114 @@
+production is checked against the mode consumption. [producer-consumer-modes] details the
+actions taken by the various possible modes of production and consumption for a Reporter value.
+
+!table caption=Actions that occur for Reporter values depending on the modes of production and
+               consumption. The prefix `REPORTER_MODE_` is omitted for clarity.
+       id=producer-consumer-modes
+| Producer Mode | Consumer Mode | Action |
```
```suggestion
!table caption=Default operations for the default context that occur for Reporter values depending on the modes of production and
               consumption. The prefix `REPORTER_MODE_` is omitted for clarity.
       id=producer-consumer-modes
| Producer Mode | Consumer Mode | Operation |
```

aeslaughter commented on framework/include/reporters/ReporterInterface.h:
```diff
@@ -0,0 +1,132 @@
+  bool hasReporterValue(const std::string & param_name) const;
+  bool hasReporterValueByName(const ReporterName & reporter_name) const;
+  ///@}
+
```
I don't understand. You can consume reporters of differing types, they just cannot have the same name.

zachmprince replied:
I'm thinking of a case where an object doesn't know explicitly what the reporter's type is, say `Real` or `int`, but the object can use a `Real` or `int` to do the same operations.

aeslaughter committed:
Apply suggestions from code review

Co-authored-by: Zachary Prince <zachmprince@gmail.com>

aeslaughter commented on framework/doc/content/syntax/Reporters/index.md:
```diff
@@ -0,0 +1,114 @@
+!table caption=Actions that occur for Reporter values depending on the modes of production and
+               consumption. The prefix `REPORTER_MODE_` is omitted for clarity.
+       id=producer-consumer-modes
+| Producer Mode | Consumer Mode | Action |
```
I used "operation" and updated the caption to use "default context"

aeslaughter commented on test/tests/reporters/mesh_info/gold/mesh_info_out.json.1:
```diff
@@ -0,0 +1,138 @@
+{
```
This shows up when you have a report data that is distributed. This is the data for rank 1

aeslaughter replied:
Here is a bit more detail. If you have a Reporter that declares a value with REPORTER_MODE_DISTRIBUTED and you run with MPI. You get a file for each rank with the distributed values.

zachmprince replied:
Okay I understand not.

But I think it should not be put in gold since you aren't using it for the test, are you?

aeslaughter commented:
I will address your other comments and squash them all up.

aeslaughter committed:
Improve split utils with max count and add reverse split

aeslaughter committed:
Use / for ReporterName and add docs about is use in validParams

(refs #11323)

aeslaughter commented:
I should probably create declareValue and declareValueByName to avoid future confusion, see #15957

aeslaughter committed:
Add ByName option for Reporter value declare method

(refs #15957)

(refs #11323)

aeslaughter committed:
Add documentation for declareValueByName

(refs #11323)

aeslaughter commented:
@dschwen FYI, The last two commits adds and document declareValueByName

zachmprince commented on framework/include/reporters/IterationInfo.h:
```diff
@@ -0,0 +1,54 @@
+
+template <typename T>
+T &
+IterationInfo::declareHelper(const std::string & item_name, T & _dummy, bool extra_check)
+{
+  return (extra_check && (!_items.isValid() || _items.contains(item_name)))
+             ? declareValueByName<T>(item_name, REPORTER_MODE_REPLICATED)
+             : _dummy;
+}
```
```suggestion
IterationInfo::declareHelper(const std::string & item_name, T & dummy, bool extra_check)
{
  return (extra_check && (!_items.isValid() || _items.contains(item_name)))
             ? declareValueByName<T>(item_name, REPORTER_MODE_REPLICATED)
             : dummy;
}
```

zachmprince reviewed (dismissed):
Couple small things. Then good to go.

aeslaughter committed:
Address reviewer comments

Test the MeshInfo json file in parallel and fix function variable name

(refs #11323)

loganharbour commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -792,13 +792,22 @@ class FEProblemBase : public SubProblem, public Restartable
-   * Initializes the postprocessor data
-   * @see SetupPostprocessorDataAction
+   * Add a Reporter object to the simulation.
+   * @param name A uniquely identifying object name
```
Need `type` for docstring

loganharbour commented on framework/include/reporters/GeneralReporter.h:
```diff
@@ -0,0 +1,26 @@
+  GeneralReporter(const InputParameters & parameters);
+
+  // These objects are not threaded
+  virtual void threadJoin(const UserObject &) final {}
```
```suggestion
  void threadJoin(const UserObject &) final {}
```

loganharbour commented on framework/include/reporters/Reporter.h:
```diff
@@ -0,0 +1,143 @@
+  const std::string & _reporter_name;
+
+  /// Needed for access to FEProblemBase::getReporterData
+  FEProblemBase * _reporter_fe_problem;
```
Can't seem to find a reason as to why this shouldn't be a reference

loganharbour commented on framework/include/reporters/ReporterContext.h:
```diff
@@ -0,0 +1,424 @@
+class ReporterContext : public ReporterContextBase
+{
+public:
+  enum class AutoOperation
```
Docstring

loganharbour commented on framework/include/reporters/ReporterData.h:
```diff
@@ -0,0 +1,319 @@
+  /// This container must be consistently sorted across processors to achieve a consistent iteration
+  /// order for parallel operations as well as for output. Objects are inserted based
+  /// on the combined ReporterName. A std::list was selected to achieve fast insert times
+  /// without re-allocating. A std::vector can also be used, but this was more fun. Do not use a
+  /// std::set, like I did initially, since the order is actually differs in each rank because it
+  /// is sorted by the pointer which causes all sorts of problems.
```
🤣

loganharbour replied:
I believe here you could actually use a std::map with a custom sorting method on `ReporterContextBase`.

This could simplify all of your getters to use the standard map find instead of doing a linear search through a vector.

loganharbour commented on framework/include/reporters/ReporterMode.h:
```diff
@@ -0,0 +1,83 @@
+  ReporterMode(const std::string & key);
+
+private:
+  // Automatically incremented in construction
+  static int ID_COUNTER;
```
Smart 👍🏼

loganharbour commented on framework/include/restart/RestartableData.h:
```diff
@@ -90,7 +98,7 @@ class RestartableData : public RestartableDataValue
    */
   T & get() { return *_value_ptr; }
-  // const T & get() const { return *_value_ptr; } // TODO: This should be used; above deprecated
+  const T & get() const { return *_value_ptr; } // TODO: This should be used; above deprecated
```
Should probably add an issue to go through and move code over if possible

aeslaughter replied:
I added a deprecated message and updated MOOSE to use the correct methods.

loganharbour commented on framework/include/utils/MooseUtils.h:
```diff
@@ -96,12 +96,17 @@ void escape(std::string & str);
  *       not. For example, "foo;bar;;" becomes {"foo", "bar", "", ""}.
  */
-std::vector<std::string> split(const std::string & str, const std::string & delimiter);
+std::vector<std::string> split(const std::string & str,
+                               const std::string & delimiter,
+                               std::size_t max_count = std::numeric_limits<std::size_t>::max());
+std::vector<std::string> rsplit(const std::string & str,
+                                const std::string & delimiter,
+                                std::size_t max_count = std::numeric_limits<std::size_t>::max());
```
But what about quotations?!

loganharbour commented on framework/src/interfaces/PostprocessorInterface.C:
```diff
@@ -20,126 +21,48 @@ PostprocessorInterface::PostprocessorInterface(const MooseObject * moose_object)
-    return _pi_feproblem.getPostprocessorValueOlder(
-        _ppi_params.get<std::vector<PostprocessorName>>(name)[index]);
-  }
+  return getPostprocessorValueHelper(name, index, 2);
```
❤️

loganharbour commented on framework/include/reporters/IterationInfo.h:
```diff
@@ -0,0 +1,54 @@
+  Real _dummy_real = 0;
+  unsigned int _dummy_unsigned_int = 0;
+
+  // Helper to perform optional declaration based on "_items"
+  template <typename T>
+  T & declareHelper(const std::string & item_name, T & _dummy, bool extra_check = true);
```
Needs a more useful docstring for now. I would put in there our intentions for moving all of the default constructable objects into the base class.

Make an issue!

aeslaughter committed:
Address review comments and updates for rebase

(refs #11323)

aeslaughter committed:
Use an access key class for getting writable ref to ReporterData

(refs #11323)

loganharbour reviewed (approved):
In terms of addressing my comments: 👍🏼
